urllib3/util/selectors.py

```diff
@@ -535,19 +535,30 @@ def close(self):
             super(KqueueSelector, self).close()
 
 
+if not hasattr(select, 'select'):  # Platform-specific: AppEngine
+    HAS_SELECT = False
+
+
 # Choose the best implementation, roughly:
 # kqueue == epoll > poll > select. Devpoll not supported. (See above)
 # select() also can't accept a FD > FD_SETSIZE (usually around 1024)
-if 'KqueueSelector' in globals():  # Platform-specific: Mac OS and BSD
-    DefaultSelector = KqueueSelector
-elif 'EpollSelector' in globals():  # Platform-specific: Linux
-    DefaultSelector = EpollSelector
-elif 'PollSelector' in globals():  # Platform-specific: Linux
-    DefaultSelector = PollSelector
-elif 'SelectSelector' in globals():  # Platform-specific: Windows
-    DefaultSelector = SelectSelector
-else:  # Platform-specific: AppEngine
-    def no_selector(_):
-        raise ValueError("Platform does not have a selector")
-    DefaultSelector = no_selector
-    HAS_SELECT = False
+def _default_selector(*_):
+    """ This function serves as a first call for DefaultSelector to
+    detect if the select module is being monkey-patched incorrectly
+    by eventlet, greenlet, and preserve proper behavior. """
+    global DefaultSelector
+    if hasattr(select, 'kqueue'):  # Platform-specific: Mac OS
+        selector_type = KqueueSelector
+    elif hasattr(select, 'epoll'):  # Platform-specific: Linux
+        selector_type = EpollSelector
+    elif hasattr(select, 'poll'):  # Platform-specific: Linux
+        selector_type = PollSelector
+    elif hasattr(select, 'select'):  # Platform-specific: Windows
+        selector_type = SelectSelector
+    else:  # Platform-specific: AppEngine
+        raise ValueError('Platform does not have a selector')
+    DefaultSelector = selector_type
```

Comment:
I don't think so, with this script and having add `print('calling _default_selector')` in the `_default_selector` function:

```
import urllib3.util.selectors

from urllib3 import PoolManager
connection = PoolManager()
print('DefaultSelector', urllib3.util.selectors.DefaultSelector)

import gevent.monkey
gevent.monkey.patch_all()

connection.urlopen('GET', 'http://google.fr')
print('DefaultSelector', urllib3.util.selectors.DefaultSelector)
connection.urlopen('GET', 'http://google.fr')
print('DefaultSelector', urllib3.util.selectors.DefaultSelector)
connection.urlopen('GET', 'http://google.fr')
print('DefaultSelector', urllib3.util.selectors.DefaultSelector)
```

I get this output:

```
('DefaultSelector', <function _default_selector at 0x101700ed8>)
('DefaultSelector', <function _default_selector at 0x101700ed8>)
calling _default_selector
calling _default_selector
('DefaultSelector', <class 'urllib3.util.selectors.SelectSelector'>)
calling _default_selector
calling _default_selector
('DefaultSelector', <class 'urllib3.util.selectors.SelectSelector'>)
```

As wait.py https://github.com/shazow/urllib3/blob/master/urllib3/util/wait.py#L3 use directly the DefaultSelector reference, it isn't replaced when setting the global `DefaultSelector`. Checkout this slides https://gitpitch.com/Lothiraldan/python-production-monkey-patching/master?grs=github&t=sky for more explanation (6 slides to the right sorry couldn't get the right sharing url).